playjson/src/main/scala/io/finch/playjson/package.scala

```diff
@@ -11,10 +13,11 @@ package object playjson {
    * @tparam A the type of the data to decode into
    */
   implicit def decodePlayJson[A](implicit reads: Reads[A]): Decode.Json[A] =
-    // TODO: Eliminate toString conversion
-    // See https://github.com/finagle/finch/issues/511
-    // PlayJson can parse from byte[] automatically detecting the charset.
-    Decode.json((b, cs) => Try(Json.parse(BufText.extract(b, cs)).as[A]))
+    Decode.json({ (b, cs) => Try(Json.parse(Buf.ByteArray.Shared.extract(b)).as[A])
+      val buf = ChannelBufferBuf.Owned.extract(b)
+      if (buf.hasArray) Try(Json.parse(new String(buf.array(), 0, buf.readableBytes(), cs)).as[A])
+      else Try(Json.parse(buf.toString(cs)).as[A])
+    })
```

Comment:
No accident. Its based on this [trick](https://github.com/finagle/finch/blob/master/core/src/main/scala/io/finch/Endpoints.scala#L287-L292) that @vkostyukov mentioned in earlier. There may be a better to express this (i.e. only having on call to `Json.parse` once and having the `if`/ `else` only return a string).